### PR TITLE
[BUG][GO] use value receiver for JSON marshal

### PR DESCRIPTION
**Pull request by perhallgren**
In the case of a required anyOf property, JSON marshalling would has been incorrect.

Required properties are not nullable, and thus always use value receivers. For the single case of anyOf models, a pointer receiver was used for MarshalJSON. All other instances of json marshalling use value receivers.

This change is simply to use a value receiver instead of a pointer receiver in the template for `MarshalJSON` on anyOf models.

#### Testing

On master, we can check that this is the only occurance of pointer receiver for json marshalling with
```bash
$ grep -ie "func (\w\+ \*.*) MarshalJSON" ./modules/openapi-generator/src/main/resources/go/*
./modules/openapi-generator/src/main/resources/go/model_anyof.mustache:func (src *{{classname}}) MarshalJSON() ([]byte, error) {
```
The above command gives an empty output with the proposed change.

Also verified with the test in [this gist](https://gist.github.com/perhallgren/efa800935a6ed520f72a3c541699d2ab). In the created tmp folder:

```bash
$ java -jar ~/code/OpenAPITools/openapi-generator/modules/openapi-generator-cli/target/openapi-generator-cli.jar generate \
        --input-spec go-json-marshal.yaml \
        --generator-name go \
        --additional-properties "withGoMod=false" \
        --package-name gen \
        --output gen
[main] INFO  o.o.codegen.DefaultGenerator - Generating with dryRun=false
[...]
$ go test .
ok  	github.com/OpenAPITools/bugs	0.004s
```

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming `7.x.0` minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by perhallgren** [comment]
Pings for technical committee: @antihax @grokify @kemokemo @jirikuncar @ph4r5h4d @lwj5

**Comment by daisukixci** [comment]
Fixed worked on my spec, thank you for it!

**Review by daisukixci — APPROVED** [review]

File: modules/openapi-generator/src/main/resources/go/model_anyof.mustache [hunk 1]
```diff
@@ -68,7 +68,7 @@ func (dst *{{classname}}) UnmarshalJSON(data []byte) error {
 }
 
 // Marshal data from the first non-nil pointers in the struct to JSON
-func (src *{{classname}}) MarshalJSON() ([]byte, error) {
+func (src {{classname}}) MarshalJSON() ([]byte, error) {
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/go/model_anyof.mustache:71]: thanks for the PR

do we need to apply the same fix model_oneof.mustache (oneOf template) as well?

Comment (perhallgren) [reply modules/openapi-generator/src/main/resources/go/model_anyof.mustache:71]: No, it already uses a value receiver. I searched for other occurances using this:

```
$ grep -ie "func (\w\+ \*.*) MarshalJSON" ./modules/openapi-generator/src/main/resources/go/*
```

Which only gives a hit in `model_anyof.mustache`.

Comment (wing328) [reply modules/openapi-generator/src/main/resources/go/model_anyof.mustache:71]: thanks for the confirmation.

can you please add a test or 2 in `samples/openapi3/client/petstore/go/model_test.go`?

you may want to update `modules/openapi-generator/src/test/resources/3_0/go/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml` with new test schemas (e.g. anyOf)

**Comment by perhallgren** [comment]
Added a test, looks OK!

Failing [here](https://app.circleci.com/pipelines/github/OpenAPITools/openapi-generator/32533/workflows/8f7c8c57-342e-4426-a2d2-1e203323ee58/jobs/106829) with:
```
    model_test.go:82: 
                Error Trace:    /home/circleci/OpenAPITools/openapi-generator/samples/openapi3/client/petstore/go/model_test.go:82
                Error:          Not equal: 
                                expected: "{\"fruit\":{\"Apple\":null,\"Banana\":{\"lengthCm\":23.4}}}"
                                actual  : "{\"fruit\":{\"lengthCm\":23.4}}"
```

And green with the fix.

**Pull request merged**